doc/zmq_getsockopt.txt

```diff
@@ -551,6 +551,14 @@ blocking or dropping sent messages. Refer to the individual socket descriptions
 in linkzmq:zmq_socket[3] for details on the exact action taken for each socket
 type.
 
+The 'ZMQ_RCVHWM' and 'ZMQ_SNDHWM' options are logical extensions to one
+another. In general, when a socket's receive high water mark has reached its
+limit and enters an exceptional state, 1) the socket will stop picking up
```

Comment:
I don't think this is correct for all socket types and configurations. A general description would need to refer to "exceptional state" only. But I am unsure if this would lead to an understandable description. The other option were to provide individual explanations for each specific behaviour in exceptional state.

Reply:
My understanding, and having confirmed through a ticket I had opened and other tickets I referenced, this behavior at least true for the TCP transport and I am fairly confident it would apply for all transports. My understanding is socket types and socket options impact only the actions taken once the high water marks are reached.

The intent here is to make clear that once a receive high watermark is reached, messages are not disregarded by the sender, but rather it "spills over" into the sender's outbound queue until it too is full. With the current description, it is easy to mistake, as you can see in the reference tickets, that messages might be disregarded if the receive high watermark is reached.

Reply:
I was not thinking about the transport. However, as you mention it, at least inproc seems to be somewhat special with regard to HWM handling, or at least for the handling of the exact parameter values.

But there are socket types such as router, which will drop messages when the exceptional state is entered. In my understanding, this is not covered by your proposed explanation.

Which socket type did you have in mind when writing this?

Moreover, this description overlaps with the docs under zmq_socket. There should be no redundancy, especially for this topic which is admittedly quite tricky.

Reply:
Dropping messages or blocking when receiving/sending them is **what** action is taken **when** the socket enters the exceptional state. What I am trying to describe here is **how** the socket ends up entering the exceptional state to begin with. **What** happens is specific to the socket type and options like you point out, but that is not what I am trying to expand on here.

The documentation right now says the exceptional state is entered when the high water marks are reached. I am trying to expand on that explanation to make clear that the the high water marks are related.

Right now, it is easy to confuse that if Socket A has 1000 messages as a receive high water mark every message from 1001 onward attempted to be sent to it by Socket B as a connected peer will be dropped until at least one of those 1000 messages are read by the application. This doesn't happen.

Socket B will buffer on its end until either Socket A notifies it "I have room for you to end those messages now" or Socket B itself has no more room to buffer because of its send high water mark and itself enters an exceptional state, where the action is thus specific to its type and options applied to it. This "cooperative" buffering doesn't seem to be specific to a socket type.

I haven't seen this explanation on the zmq_socket docs page either. The explanations there simply describe **what** the action is when the exceptional state is entered. If I am mistaken, let me know what I missed and I can take a look again.